website/blog/2024-02-05-running-tabby-as-a-language-server/index.md

```diff
@@ -0,0 +1,85 @@
+
+## What is Tabby Agent
+
+[Tabby Agent](https://github.com/tabbyml/tabby/tree/main/clients/tabby-agent/) is a Node.js package that communicates with the Tabby server and implements several useful features for code completion, such as:
+
+- **Debouncing**: Tabby Agent handles code completion requests by implementing an appropriate debouncing mechanism. This ensures that requests are not sent too frequently, reducing server load and improving performance, as the automatic inline code completion often listens for text input which can be very frequent.
+- **Caching**: To prevent redundant completion requests, KV caching is implemented in Tabby Agent. When a completion is dismissed but needs to be requested again at the same location, the cached content is used directly. Cached completions are also matched when the prefix of a request aligns with a previously cached completion, eliminating the need for additional server requests. This is especially useful when the user types the same text with the ghost text suggestions.
+- **Post-processing**: Tabby Agent enhances the completion results through post-processing. Tabby Agent filters out low-quality completions, improving the overall user experience by removing duplicate suggestions. Additionally, it limits the length of suggestions to the focused scope. This helps users focus on the most relevant suggestions.
+
+These features were initially developed as part of the [Tabby VSCode extension](https://github.com/tabbyml/tabby/tree/main/clients/vscode/). However, we wanted to reuse this common client code logic and avoid reimplementing it for other text editors. As a result, Tabby Agent was created as a standalone Node.js package that can be used by other editors to communicate with Tabby.
+
+![tabby-agent](./tabby-agent.png)
+
+## Why Language Server
+
+Tabby Agent previously utilized a customized protocol based on JSON Lines, initially designed for compatibility with VIM's JSON mode channel. However, this protocol was not widely used, which increased the difficulty of integrating Tabby Agent with different editors. By adopting a more universal protocol, we can offer a more flexible and streamlined experience in creating Tabby plugins for various editors.
+
+The [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) defines a standardized protocol for communication between a language server and its clients. It provides methods that can be used to implement a wide range of features, including code completion.
+
+![tabby-agent-lsp](./tabby-agent-lsp.png)
+
+Running Tabby as Language Server will provides code completion functionality using the standard `textDocument/completion` protocol. It can suggest code completions based on the context of the code, whether it's a line or a block, rather than just a single word.
```

Comment:
```suggestion
Running Tabby as a Language Server provides code completion with the standard `textDocument/completion` protocol. It can suggest code completions based on the context of the code, whether it's a line or a block, rather than just a single word.
```